### ISSUE TITLE
useStore stops firing on state changes after component remounts

### ISSUE THREAD

**Issue by karlpetersson** [react-spring/zustand#86]
When switching between mounting two components that both calls the useStore hook, where one of them conditionally mounts a child that also calls useStore, the parent will stop rendering on state changes after remounting. Here's a minimal example on codesandbox: 

https://codesandbox.io/s/nostalgic-hooks-0x63r

Repro steps are included. Let me know if you want the code pasted here instead.

Could this be related to #85 ?

edit: updated codesandbox link, messed around in the original one I posted without realizing it.

**Comment by karlpetersson** [comment]
Downgrading to 2.2.1 seems to fix this issue. Tested by downgrading the zustand package in the above codesandbox.

**Comment by karlpetersson** [comment]
Might be a dumb question, but why not make the array dense right after adding the subscriber? This seems to solve this issue as well as work with #84, while keeping indices intact. E.g:

```
const subscribe: Subscribe<TState> = <StateSlice>(
    subscriber: Subscriber<TState, StateSlice>
  ) => {
    // ...
    subscribers[subscriber.index] = subscriber;

    subscriberCount = 0;
    subscribers = subscribers.filter(s => {
      subscriber.index = subscriberCount++;
      return true;
    });

    return () => delete subscribers[subscriber.index];
  };
```

**Comment by Jamikk** [comment]
I have the same problem

**Comment by fzkirablackwhy** [comment]
+

**Comment by fzkirablackwhy** [comment]
me too

**Comment by lolk12** [comment]
me too! HELP ZUSTAND

**Comment by dentrado** [comment]
Any news on this? Seems to still be a problem

**Comment by lolk12** [comment]
@dentrado On version 2.2.1, everything is fine

**Comment by dentrado** [comment]
But then I get the bug they fixed in 2.2.2 instead

**Comment by lolk12** [comment]
@dentrado very sadly (
